# About Pascal’s Triangle
## Class Directed Learning
### Due: Tuesday, 2 February 2021, 11:59 pm

## A Helper Function
Exercise 253 on page 161 shows a recursive definition of `nCk`. Here's a much more efficient way to compute binomial coefficients, which perhaps you can profitably use in your investigations:


In [4]:
from math import gcd

def nCk(n, k):
  if k < 0 or k > n:
    return 0
  else:
    result = 1
    d = 1
    g = 1
    m = min(k, n - k)
    while (d <= m):
      g = gcd(result, d)
      result = n * (result // g)
      result = (result // (d // g))
      n -= 1
      d += 1
    return result

In [9]:
for i in range(0, 20):
  for j in range(0, i+1):
    print(nCk(i,j), end=' ')
  print()

1 
1 1 
1 2 1 
1 3 3 1 
1 4 6 4 1 
1 5 10 10 5 1 
1 6 15 20 15 6 1 
1 7 21 35 35 21 7 1 
1 8 28 56 70 56 28 8 1 
1 9 36 84 126 126 84 36 9 1 
1 10 45 120 210 252 210 120 45 10 1 
1 11 55 165 330 462 462 330 165 55 11 1 
1 12 66 220 495 792 924 792 495 220 66 12 1 
1 13 78 286 715 1287 1716 1716 1287 715 286 78 13 1 
1 14 91 364 1001 2002 3003 3432 3003 2002 1001 364 91 14 1 
1 15 105 455 1365 3003 5005 6435 6435 5005 3003 1365 455 105 15 1 
1 16 120 560 1820 4368 8008 11440 12870 11440 8008 4368 1820 560 120 16 1 
1 17 136 680 2380 6188 12376 19448 24310 24310 19448 12376 6188 2380 680 136 17 1 
1 18 153 816 3060 8568 18564 31824 43758 48620 43758 31824 18564 8568 3060 816 153 18 1 
1 19 171 969 3876 11628 27132 50388 75582 92378 92378 75582 50388 27132 11628 3876 969 171 19 1 


## Predominant and Fundamental
The predominant pattern, the fundamental symmetry in Pascal's triangle is captured by the identity:
$${n \choose k} = {n \choose n - k}$$
## Identify Identities
Here are two different identities that point out the same pattern in Pascal's triangle. What is the name of this pattern?
### Identity 1
For nonnegative integers $r$ and $n$:
$$\sum^n_{k=0} {r+k \choose k} = {r+n+1 \choose n}$$
### Identity 2
For nonnegative integers $m$ and $n$:
$$\sum^n_{k=0} {k \choose m} = {n+1 \choose m+1}$$

# Investigate A Possible Relationship
Does any row in Pascal's Triangle have the Fibonacci numbers' relationship between three consecutive entries? In other words, where x, y, and z are three consecutive entries in a single row, does x + y = z?

  The number of rows where this relationship could possibly hold is:
  - [x] Zero.
  - [ ] At least one.
  - [ ] At most one.
  - [ ] Exactly one.
  - [ ] It cannot be determined.

# Investigate Another Possible Relationship
Does any row in Pascal's Triangle have three consecutive entries in the ratio 1:2:3? (12:24:36 is an example of three numbers in a 1:2:3 ratio.)?

  The number of rows where this relationship could possibly hold is:
  - [ ] Zero.
  - [x] At least one.
  - [ ] At most one.
  - [ ] Exactly one.
  - [ ] It cannot be determined.

1001 2002 3003 is a sequence of the 1:2:3 ratio that occurs in line 15 of the pascal triangle